lib/spack/spack/environment.py

```diff
@@ -337,47 +395,62 @@ def __init__(self, path, init_file=None):
+                path to the view.
         """
         self.path = os.path.abspath(path)
         self.clear()
 
         if init_file:
-            # initialize the environment from a file if provided
             with fs.open_if_filename(init_file) as f:
                 if hasattr(f, 'name') and f.name.endswith('.lock'):
-                    # Initialize the environment from a lockfile
+                    self._read_manifest(default_manifest_yaml, init=True)
                     self._read_lockfile(f)
                     self._set_user_specs_from_lockfile()
-                    self.yaml = _read_yaml(default_manifest_yaml)
                 else:
-                    # Initialize the environment from a spack.yaml file
-                    self._read_manifest(f)
+                    self._read_manifest(f, init=True)
         else:
-            # read lockfile, if it exists
+            init = not any(os.path.exists(x)
+                           for x in (self.lock_path, self.manifest_path))
+            default_manifest = not os.path.exists(self.manifest_path)
+            if default_manifest:
+                self._read_manifest(default_manifest_yaml, init=init)
+            else:
+                with open(self.manifest_path) as f:
+                    self._read_manifest(f, init=init)
+
             if os.path.exists(self.lock_path):
                 with open(self.lock_path) as f:
                     self._read_lockfile(f)
+                if default_manifest:
+                    self._set_user_specs_from_lockfile()
 
-            if os.path.exists(self.manifest_path):
-                # read the spack.yaml file, if exists
-                with open(self.manifest_path) as f:
-                    self._read_manifest(f)
-
-            elif self.concretized_user_specs:
-                # if not, take user specs from the lockfile
-                self._set_user_specs_from_lockfile()
-                self.yaml = _read_yaml(default_manifest_yaml)
-            else:
-                # if there's no manifest or lockfile, use the default
-                self._read_manifest(default_manifest_yaml)
+        if with_view is False:
+            self._view_path = None
+        elif isinstance(with_view, six.string_types):
+            self._view_path = with_view
+        # If with_view is None, then defer to the view settings determined by
+        # the manifest file
 
-    def _read_manifest(self, f):
+    def _read_manifest(self, f, init=False):
         """Read manifest file and set up user specs."""
         self.yaml = _read_yaml(f)
         spec_list = config_dict(self.yaml).get('specs')
         if spec_list:
             self.user_specs = [Spec(s) for s in spec_list if s]
 
+        enable_view = config_dict(self.yaml).get('view')
+        # enable_view can be true/false, a string, or None (if the manifest did
+        # not specify it)
+        if enable_view is True or (enable_view is None and init):
```

Comment:
If we _really_  need to pass in `init` for this logic I think maybe this chunk of code belongs in another function with a more descriptive name, that should just be called immediately after `read_manifest` by code in `environment.py`.  Is there a reason not to do that?